fix(ivy): attempting to remove detached view on destroy

crisbeto opened the pull request:
Currently the `ViewRef.destroy` method assumes that its index inside the view container will always be valid, however if it has been removed already, it'll be -1 which will throw an error.

The error manifested itself in one of the unit tests where a view had been detached during the test and then `TestBed` attempted to destroy its `ComponentRef` which ended up throwing an `Error during cleanup of component`.

mary-poppins commented:
You can preview 5f76377 at https://pr27585-5f76377.ngbuilds.io/.

pkozlowski-opensource commented:
> The error manifested itself in one of the unit tests

@crisbeto do you mean an existing TestBed test? If so we should remove `fixmeIvy` from it in this PR. If the fix in this PR is not enough to remove  `fixmeIvy`, we should add a dedicated runtime or TestBed test. This way or another we need a test for this case :-)

crisbeto commented:
I should've worded it better. I believe it ended up throwing inside `afterEach` which is why it didn't actually fail any of the tests. Regarding a unit test, I wasn't too sure where to put one since there wasn't a dedicated file for `ViewRef`.

kara reviewed (changes requested):
Looks good, but as @pkozlowski-opensource pointed out, this needs a runtime test. Can you add one to `view_container_ref_spec.ts`?

crisbeto committed:
fix(ivy): attempting to remove detached view on destroy

Currently the `ViewRef.destroy` method assumes that its index inside the view container will always be valid, however if it has been removed already, it'll be -1 which will throw an error.

The error manifested itself in one of the unit tests where a view had been detached during the test and then `TestBed` attempted to destroy its `ComponentRef` which ended threw an `Error during cleanup of component`.

crisbeto commented:
I've added a test case @kara @pkozlowski-opensource.

mary-poppins commented:
You can preview c8ad800 at https://pr27585-c8ad800.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/src/render3/view_ref.ts:
```diff
@@ -61,7 +61,12 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
       this._appRef.detachView(this);
     } else if (this._viewContainerRef && viewAttached(this._lView)) {
-      this._viewContainerRef.detach(this._viewContainerRef.indexOf(this));
+      const index = this._viewContainerRef.indexOf(this);
```
Seems like this fix is also part of another PR, see https://github.com/angular/angular/pull/27604/files#diff-6214bef312f4279eb8c0a3f40aed7511R64

crisbeto replied:
@mhevery how should we handle this one? We had a quick talk with @kara about it yesterday and the logic needs to be changed slightly in this block, aside from the index check (the `viewAttached` call should go away as well). These changes might a bit easier to get in since they're fairly small.

mhevery replied:
Discussed with @kara and we think the correct fix is in fixup 4fdcf01aa0.

mhevery committed:
fixup! fix(ivy): attempting to remove detached view on destroy

mary-poppins commented:
You can preview 4fdcf01 at https://pr27585-4fdcf01.ngbuilds.io/.

kara reviewed (approved):
LGTM

mhevery committed:
fixup! fix(ivy): attempting to remove detached view on destroy

mary-poppins commented:
You can preview c60187e at https://pr27585-c60187e.ngbuilds.io/.
